mmdet/models/dense_heads/paa_head.py

```diff
@@ -0,0 +1,603 @@
+
+                - label (Tensor): classification target of each anchor after
+                  paa assign, with shape (num_anchors,)
+                - label_weight (Tensor): Classification loss weight of each
+                  anchor after paa assign, with shape (num_anchors).
+                - bbox_weight (Tensor): Bbox weight of each anchor with shape
+                  (num_anchors, 4).
+                - num_pos (int): The number of positive samples after paa
+                  assign.
+        """
+        if not len(pos_inds):
+            return label, label_weight, bbox_weight, 0
+
+        num_gt = pos_gt_inds.max() + 1
+        num_level = len(anchors)
+        anchors_all_level = torch.cat(anchors, 0)
+        num_anchors_each_level = [item.size(0) for item in anchors]
+        pos_scores = cls_score[pos_inds]
+        pos_bbox_pred = bbox_pred[pos_inds]
+        pos_label = label[pos_inds]
+        pos_label_weight = label_weight[pos_inds]
+        pos_bbox_target = bbox_target[pos_inds]
+        pos_bbox_weight = bbox_weight[pos_inds]
+        pos_anchors = anchors_all_level[pos_inds]
+        pos_bbox_pred = self.bbox_coder.decode(pos_anchors, pos_bbox_pred)
+
+        # to keep loss dimension
+        origin_reduction = self.loss_cls.reduction
+        self.loss_cls.reduction = 'none'
+        loss_cls = self.loss_cls(
+            pos_scores, pos_label, pos_label_weight,
+            avg_factor=None) / self.loss_cls.loss_weight
+        self.loss_cls.reduction = origin_reduction
+
+        origin_reduction = self.loss_bbox.reduction
+        self.loss_bbox.reduction = 'none'
+        loss_bbox = self.loss_bbox(
+            pos_bbox_pred, pos_bbox_target, pos_bbox_weight,
+            avg_factor=None) / self.loss_bbox.loss_weight
+        self.loss_bbox.reduction = origin_reduction
+
+        loss_cls = loss_cls.sum(-1)
+        pos_loss = loss_bbox + loss_cls
+        assert not torch.isnan(pos_loss).any()
+        inds_level_interval = [
+            sum(num_anchors_each_level[:i])
+            for i in range(0,
+                           len(num_anchors_each_level) + 1)
+        ]
+        pos_level_mask = []
+        for i in range(num_level):
+            mask = (pos_inds >= inds_level_interval[i]) & (
+                pos_inds < inds_level_interval[i + 1])
+            pos_level_mask.append(mask)
+        pos_inds_after_paa = []
+        for gt_ind in range(num_gt):
+            pos_inds_gmm = []
+            pos_loss_gmm = []
+            gt_mask = pos_gt_inds == gt_ind
+            for level in range(num_level):
+                level_mask = pos_level_mask[level]
+                level_gt_mask = level_mask & gt_mask
+                value, topk_inds = pos_loss[level_gt_mask].topk(
+                    min(level_gt_mask.sum(), self.topk), largest=False)
+                pos_inds_gmm.append(pos_inds[level_gt_mask][topk_inds])
+                pos_loss_gmm.append(value)
```

Comment:
Better do a tensor implementation for this part.

Reply:
All variables is tensor in this loop